bsp/Src/UART.c

```diff
@@ -346,22 +369,34 @@ uart_status_t uart_recv(UART_HandleTypeDef* handle, uint8_t* data, uint8_t lengt
     #endif /* UART5 */
 
     uart_status_t status = UART_RECV;
+    rx_payload_t receivedPayload;
+    uint8_t bytes_received = 0;
 
     // Receive all requested bytes
-    for (uint8_t i = 0; i < length; i++) {
-        if (xQueueReceive(rx_queue, &data[i], delay_ticks) == errQUEUE_EMPTY) {
-            return UART_EMPTY;
-        } 
+    while (bytes_received < length) {
+        if (xQueueReceive(rx_queue, &receivedPayload, delay_ticks) == errQUEUE_EMPTY) {
+            return UART_EMPTY;  // Queue empty, no more data to receive
+        }
+
+        // Calculate how many bytes to copy from the payload based on DATA_SIZE
+        uint8_t copy_length = (length - bytes_received) >= DATA_SIZE ? DATA_SIZE : (length - bytes_received);
+
+        // Copy the data from the payload to the user's data buffer
+        for (uint8_t i = 0; i < copy_length; i++) {
+            data[bytes_received + i] = receivedPayload.data[i];
+        }
+
+        // Update the number of bytes received
+        bytes_received += copy_length;
     }
-    
 
     return status;
 }
 
 // Transmit Callback occurs after a transmission if complete (depending on how huart is configure)
 void HAL_UART_TxCpltCallback(UART_HandleTypeDef *huart) {
     BaseType_t higherPriorityTaskWoken = pdFALSE;
-    uint8_t tx_buffer[32];  // Buffer for collecting bytes to send
+    uint8_t tx_buffer[32];  // Buffer for collecting bytes to send, though still unsure if this is completely safe, since it is local on the stack, may need a static buffer for each UART instance?
```

Comment:
this should be fine as long as the tx_buffer pointer doesn't escape the ISR. since ur just using it locally it's chill